pkg/cmd/codespace/codespace_selector.go

```diff
@@ -0,0 +1,138 @@
+package codespace
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/cli/cli/v2/internal/codespaces/api"
+	"github.com/spf13/cobra"
+)
+
+type CodespaceSelector interface {
+	// Select fetches a codespace from the API according to the criteria set at its creation, prompting the user if needed
+	Select(ctx context.Context) (codespace *api.Codespace, err error)
+
+	// SelectName fetches a codespace name in the same way as Select, but may avoid API calls
+	SelectName(ctx context.Context) (name string, err error)
+}
+
+type codespaceSelector struct {
+	api apiClient
+
+	repoName      string
+	codespaceName string
+}
+
+var errNoFilteredCodespaces = errors.New("you have no codespaces meeting the filter criteria")
+
+// AddCodespaceSelector add codespace filtering parameters to the give command and returns a CodespaceSelector which applies them
+func AddCodespaceSelector(cmd *cobra.Command, app *App) *codespaceSelector {
+	cs := &codespaceSelector{api: app.apiClient}
+
+	cmd.Flags().StringVarP(&cs.codespaceName, "codespace", "c", "", "Name of the codespace")
+	cmd.Flags().StringVarP(&cs.repoName, "repo", "R", "", "Filter codespace selection by repository name (user/repo)")
+
+	cmd.MarkFlagsMutuallyExclusive("codespace", "repo")
+
+	return cs
+}
+
+// AddDeprecatedRepoShorthand adds a -r parameter (deprecated shorthand for --repo)
+// which instructs the user to use -R instead.
+func (cs *codespaceSelector) AddDeprecatedRepoShorthand(cmd *cobra.Command) error {
+	cmd.Flags().StringVarP(&cs.repoName, "repo-deprecated", "r", "", "(Deprecated) Shorthand for --repo")
+
+	if err := cmd.Flags().MarkHidden("repo-deprecated"); err != nil {
+		return fmt.Errorf("error marking `-r` shorthand as hidden: %w", err)
+	}
+
+	if err := cmd.Flags().MarkShorthandDeprecated("repo-deprecated", "use `-R` instead"); err != nil {
+		return fmt.Errorf("error marking `-r` shorthand as deprecated: %w", err)
+	}
+
+	cmd.MarkFlagsMutuallyExclusive("codespace", "repo-deprecated")
+
+	return nil
+}
+
+func (cs *codespaceSelector) Select(ctx context.Context) (codespace *api.Codespace, err error) {
```

Comment:
One idea to perhaps simplify things is:
```suggestion
func (cs *codespaceSelector) Select(ctx context.Context, skipInfate bool) (codespace *api.Codespace, err error) {
```
Where `skipInflate` would indicate that the caller only needs the name of the codespace to be populated. This would allow for removing `SelectName`. I don't necessarily love this idea but it was a thought I had.

Reply:
I had considered that too, but chose not to go that route for a couple reasons. 1) Personally I'm not a big fan of returning partially-filled objects, it always feels like asking for misunderstanding or coupling. And 2) I could see future use cases of this sort of method like `SelectId` or `SelectConnectionCredentials`. At that point it may make more sense make a unified method, but I wanted to see those use case requirements first in order to craft a cleaner generic solution at that point.